modules/dag-l1/src/main/scala/org/tessellation/dag/l1/domain/block/BlockProcessor.scala

```diff
@@ -0,0 +1,104 @@
+import org.tessellation.dag.l1.domain.transaction.TransactionService
+import org.tessellation.kryo.KryoSerializer
+import org.tessellation.schema.address.Address
+import org.tessellation.schema.balance.Balance
+import org.tessellation.schema.transaction.Transaction
+import org.tessellation.security.signature.Signed
+import org.tessellation.security.{Hashed, SecurityProvider}
+
+import eu.timepit.refined.auto._
+import eu.timepit.refined.numeric.NonNegative
+import eu.timepit.refined.refineV
+import io.estatico.newtype.ops.toCoercibleIdOps
+
+class BlockProcessor[F[_]: Async: KryoSerializer: SecurityProvider](
+  addressStorage: AddressStorage[F],
+  blockService: BlockService[F],
+  blockValidator: BlockValidator[F],
+  transactionService: TransactionService[F]
+) {
+
+  def accept(signedBlock: Signed[DAGBlock]): F[Unit] =
+    for {
+      _ <- blockValidator.validate(signedBlock).map {
+        case Valid(_)         => Applicative[F].unit
+        case Invalid(reasons) => BlockFailedValidation(reasons).raiseError[F, Unit]
+      }
+
+      hashedBlock <- signedBlock.hashWithSignatureCheck.flatMap(_.liftTo[F])
+      hashedTransactions <- signedBlock.value.transactions.toList
+        .traverse(_.hashWithSignatureCheck.flatMap(_.liftTo[F]))
+        .map(_.toSet)
+
+      _ <- acceptTransactions(hashedTransactions)
+      _ <- blockService.accept(hashedBlock)
+      _ <- blockService.handleTipsUpdate(hashedBlock)
+    } yield ()
+
+  private def prepareBalances(transactions: Set[Transaction]): F[Map[Address, Balance]] = {
+    val sources = transactions.groupBy(_.source)
+    val destinations = transactions.groupBy(_.destination)
+    val all = sources.combine(destinations)
+    val addressBalanceChanges = all.map {
+      case (address, txs) =>
+        val change = txs.foldLeft(BigInt(0L))(
+          (change, tx) =>
+            tx match {
+              case Transaction(`address`, _, amount, fee, _, _) => change - amount.coerce - fee.coerce
+              case Transaction(_, `address`, amount, _, _, _)   => change + amount.coerce
+              case _                                            => change
+            }
+        )
+
+        address -> change
+    }
+
+    addressBalanceChanges.toList.traverse {
+      case (address, balanceChange) =>
+        for {
+          current <- addressStorage.getBalance(address)
+          newBalance <- refineV[NonNegative]
+            .apply(current.coerce + balanceChange)
+            .bimap(BalanceNegative, Balance(_))
+            .liftTo[F]
+        } yield (address, newBalance)
+    }.map(_.toMap)
+  }
+
+  private def acceptTransactions(hashedTransactions: Set[Hashed[Transaction]]): F[Unit] =
+    for {
+      preparedBalances <- prepareBalances(hashedTransactions.map(_.signed.value))
+      _ <- addressStorage.updateBalances(preparedBalances)
+      _ <- hashedTransactions.toList.traverse(transactionService.accept)
```

Comment:
I think that it's a responsibility of transaction service to accept balances accordingly. If you want to accept multiple txs at once, I'd create a function that receives set of txs and accepts them all or nothing. Anyway, looks like you update balances before you accept the transaction and I think it's wrong

Reply:
We would need to switch from `MapRef` to `Ref[Map[K, V]]` to be able to do an atomic update on transactions. Although it still won't give us complete atomicity unfortunately (on block acceptance level) because we store balances in a different cache (database actually) and blocks in yet different one. Theoretically any of these may fail so if that actually happens during the updates we have a problem either way. I think I intended the balances update to happen first because given that we chose to store these inside of a database (and the database seems like the most erratic of the storages we use here) there is the biggest probability that block acceptance will fail at that moment - if it does it's still ok, balances update failed and nothing else got accepted so we may fail the whole flow with no consequences. After accepting balances everything after that should succeed and the validations and synchronized acceptance of blocks should enforce that. But from the local reasoning point of view I can't guarantee that if the transactions got accepted the block will also get accepted these are side effects wrapped in `F` so they may throw. Thinking about a solution for that case (because I don't see a way to atomically change all three caches at the same time), after an unexpected error is thrown we can make the node switch to a `Misaligned` status and not perform any operations until new majority globalSnapshot is received and then if the node is Misaligned that would trigger a redownload to that new majority right of the bat - we could also eagerly redownload to the last known majority. I planned to create a ticket regarding atomic acceptance and I think it makes sense to do so and try to spike this case there.

Regarding the `transactionService` being responsible for balance update I'm not sure, after all `transactionService` doesn't store the balances so why would it be responsible for updating them. Also if we start to nest classes the solution may get less elegant. `BlockService` would need `TransactionService` which would need `AddressStorage`, so it's kind of getting a bit convoluted. I strived towards flatter structure and decoupling to avoid circular dependencies. Just my thoughts, not saying that's the way to go.

Reply:
But it means that when I take the transactionService and call “accept”, the transaction will not be fully accepted 🤔